tests/data/modules/exp/plugins/__init__.py

```diff
@@ -39,3 +40,29 @@ class Test2(resources.PurgeableResource):
     """
 
     fields = ("name", "agent", "mydict", "mylist")
+
+
+@resources.resource("exp::RequiresTest", agent="agent", id_attribute="name")
+class RequiresTest(resources.PurgeableResource):
+
+    fields = ("name", "agent", "do_break")
+
+
+@dependency_manager
+def bad_loops(_, all_resources):
+    my_resources = {}
+
+    for resource in all_resources.values():
+        if resource.id.get_entity_type() == "exp::RequiresTest":
+            my_resources[resource.id] = resource
+
+    for resource in my_resources.values():
+        # wire into loop
+        for require_id in resource.requires:
+            if require_id in my_resources:
+                required = my_resources[require_id]
+                required.requires.add(resource.id)
+        if resource.do_break == 1:
+            resource.requires.add("xyz")
```

Comment:
This violates the type definition of `Resource.requires`. I assume this is not specific to this PR. However, the too narrow type definition led to confusion about the usefulness of the type of the `cleanup` function in `export.py`. Should a ticket be created to expand the type definition?

Reply:
There are two issues here:
1- dependency managers are semi-trusted code, so they can not be trusted to respect the types. Part of this pull request is making sure they are caught when they mess with it.
2- We already have dependency managers deployed that use strings, Ids and Resources. The typing as it written now is the type we expect. If dependency managers violate the typing, this is cleaned up by the cleanup method.

Reply:
I think the best approach here is to document this? Do you agree?

Reply:
I guess that makes sense, though it makes reasoning about code harder since you need to take into account which input is accessible by semi-trusted code. But I think that's just a limitation of python so I'll resolve this.

Reply:
My comment was out of sync. It was a reply to your first response. I agree that documenting this is the best approach.

Reply:
It is even worse: all code is accessible to semi-trusted code. But this specific code is intended to be accessed by it.